src/test/java/co/cask/hydrator/salesforce/etl/SalesforceBatchSourceETLTest.java

```diff
@@ -0,0 +1,194 @@
+
+    Assert.assertEquals(sObjects.size(), records.size());
+
+    StructuredRecord record = records.get(0);
+
+    Assert.assertEquals("Proposal", (String) record.get("StageName"));
+    Assert.assertEquals(false, (boolean) record.get("IsDeleted"));
+    Assert.assertEquals(25.0, (double) record.get("TotalOpportunityQuantity"), 0.01);
+    Assert.assertEquals(System.currentTimeMillis(), (long) record.get("LastModifiedDate"),
+                        TimeUnit.MINUTES.toMillis(30));
+  }
+
+  @Test
+  public void testLikeClause() throws Exception {
+    List<SObject> sObjects = new ImmutableList.Builder<SObject>()
+      .add(new SObjectBuilder()
+             .setType("Account")
+             .put("Name", "testLikeClause1")
+             .build()
+      )
+      .add(new SObjectBuilder()
+             .setType("Account")
+             .put("Name", "testLikeClause2")
+             .build()
+      )
+      .add(new SObjectBuilder()
+             .setType("Account")
+             .put("Name", "testLikeClause3")
+             .build()
+      )
+      .build();
+
+    addSObjects(sObjects);
+
+    String query = "SELECT Id, Name FROM Account WHERE Name LIKE 'testLikeClause%'";
+    List<StructuredRecord> records = getResultsBySOQLQuery(query);
+
+    Assert.assertEquals(sObjects.size(), records.size());
+
+    Set<String> names = new HashSet<>();
+    names.add(records.get(0).get("Name"));
+    names.add(records.get(1).get("Name"));
+    names.add(records.get(2).get("Name"));
+
+    Set<String> expectedNames = new HashSet<>();
+    expectedNames.add("testLikeClause1");
+    expectedNames.add("testLikeClause2");
+    expectedNames.add("testLikeClause3");
+    Assert.assertEquals(expectedNames, names);
+  }
+
+  @Test
+  public void testWhereClause() throws Exception {
+    // Login entry will be created automatically no need to create it
+    String query = "SELECT UserId from LoginHistory WHERE LoginTime > " +
+      "2000-09-20T22:16:30.000Z AND LoginTime < 9999-09-21T22:16:30.000Z";
+    List<StructuredRecord> records = getResultsBySOQLQuery(query);
+    Assert.assertNotEquals(0, records.size());
+
+    StructuredRecord record = records.get(0);
+    Assert.assertFalse(((String) record.get("UserId")).isEmpty());
+  }
+
+  @Test
+  public void testEmptyQueryResponse() throws Exception {
+    String query = "SELECT Id from Opportunity LIMIT 0";
+    List<StructuredRecord> records = getResultsBySOQLQuery(query);
+    Assert.assertEquals(0, records.size());
+  }
+
+  @Test
+  public void testSOQLSyntaxError() throws Exception {
+    String query = "SELECTS something FROM table";
+    try {
+      getResultsBySOQLQuery(query);
+      Assert.fail("Expected syntax error");
+    } catch (IllegalStateException e) {
+      /*
+      java.lang.IllegalStateException: Expected 200 OK, got 400 Bad Request. Error: Failed to deploy app
+
+
+      This is exception from workflow manager, not the actual exception with syntax error.
```

Comment:
is there any way to unwrap the exception thrown? seems bad if the original exception is not propagated

Reply:
This exception is thrown by startAndWaitForRun which monitors cdap task via API. The exception is  thrown inside actual pipeline run separately. So I am afraid there is no simple way to get it.

Reply:
sg